### DISCUSSION TITLE
Is there a way omit keys from a discriminated union? Or an alternative way to achieve something similar?

### DISCUSSION THREAD

**Discussion by kculmback-eig** [colinhacks/zod#1434] [Q&A] [answered]
We're hoping to do something like this:

```typescript
export const BaseTemplateSchema = z.object({
  id: z.string(),
  name: z.string(),
  lastUpdated: z.string().optional(),
  createdOn: z.string(),
})

export const CommunicationTemplateSchema = BaseTemplateSchema.merge(
  z.object({
    type: z.literal('templates:types:communication'),
    smsTemplate: z.string().optional().nullable(),
    subject: z.string(),
    htmlTemplate: z.string(),
    textTemplate: z.string(),
  })
)

export const NoteTemplateSchema = BaseTemplateSchema.merge(
  z.object({
    type: z.literal('templates:types:note'),
    textTemplate: z.string(),
  })
)

export const TemplateSchema = z.discriminatedUnion('type', [
  CommunicationTemplateSchema,
  NoteTemplateSchema,
])

export const CreateTemplateSchema = TemplateSchema.omit({
  id: true,
  createdOn: true,
  lastUpdated: true,
})
```

This doesn't work when we get to `CreateTemplateSchema` because `omit` does not exist on the `ZodDiscriminatedUnion` type.

Essentially what we're trying to do is to have a TemplateSchema type that has different keys depending on the `type` field. But we also want to be able to omit certain keys from the type for certain requests (e.g. you can't provide an id when creating a template).

Is there a way to achieve this with zod? Wether using discriminated unions or something else?

**Comment by scotttrinh** [comment, marked as answer]
Yeah, we've talked previously about adding the other `ZodObject` methods to `ZodDiscriminatedUnion`, but haven't made any progress on that that I'm aware of. The temporary fix is to make a separate union like this:

[TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwF4F84DMoRHAREiAE1wG4AoMgUwA9JY4BjCAOwGd4AhAQ1coBVK4ADZcYlAMoMAFoK5wAvHCQA6CACMAVpQYwAFAjJw4wQgC4ly9lGDMA5roCUAGkNxmXEJXMqrN+89cRdgBVMEJRSjMLXztHVTAYYBYuIUcXIwYoSgjCAHlmb0sYa1iAlAcKGjp4JjZ4AGFsEABXZmAGUSTmAWEIyRkQOUVuXh6wETF+2WVPKFtKXVcVdS0dfVcjGABPMC8LIWAxKBTdAHIxXrFWU23d66YQFraOxJZTgKMjVhBWMYm9nzFPxxCAJLonBzKZjNIQiNRCBYfT6sZqabQwQoxfzpT5SGAgIR-CKYoGlHGbGgwIliEklbGucpkCpUWjQGosdhwAByEDE1Ik0lkCjgI34gnGfUFgxmlDmCyWqjRawMnzgtwBygORxO53F-2u6uuzF5lHe5LVlP5tOBSMZzKqbMYHPg-Kmg2FKkIwFYmVANhywTaLDO6tOTjgAG1XI1Hq12p0WK6pVwcTy+XrJQMU2QALrM2qczLZMS5ECHVjClXGKLFZqUHFFnL5cy1+uBHgwULhMQ1qB1lxoHhOurkFnVYec+pZCJJrMe5Ren3WMvuHuBrq6XDq3DhqNGGNPeOvboZybJ1RlvSNkuX1hItNiy4CrMXw66a+UUvlgJ5khAA)

```ts
const createOmits = {
  id: true,
  createdOn: true,
  lastUpdated: true,
} as const;

export const CreateTemplateSchema = z.discriminatedUnion("type", [
  CommunicationTemplateSchema.omit(createOmits),
  NoteTemplateSchema.omit(createOmits),
]);
```

**Reply by kculmback-eig** [reply]
Okay, thank you!

**Comment by wiese** [comment]
Somewhat related: I needed Omit for the type created by `z.infer<typeof myDiscriminatedUnion>;` and got the removal of some properties working using https://github.com/microsoft/TypeScript/issues/31501#issuecomment-1079728677

**Comment by johnsvanda** [comment]
Any update for this?

**Comment by lukachi** [comment]
```typescript
const MySchema = z.discriminatedUnion('type', [
  z.object({ type: z.literal('example1'), foo: z.string() }),
  z.object({ type: z.literal('example2'), bar: z.number() }),
])

const myNewTypeWithoutEx2: Exclude<
  z.infer<typeof MySchema>,
  { type: 'example2' }
>
```

**Reply by amatiasq** [reply]
That solves a different problem.
